backend/python/tapa/add_axi_pipeline.py

```diff
@@ -0,0 +1,219 @@
+import pyverilog.vparser.ast as ast
+import re
+
+from pyverilog.vparser.parser import parse as rtl_parse
+from pyverilog.vparser.ast import Input, Output, Parameter
+from pyverilog.ast_code_generator.codegen import ASTCodeGenerator
+from typing import *
+
+
+class AXI:
+  def __init__(self, name, data_width, addr_width):
+    self.name = name
+    self.data_width = data_width
+    self.addr_width = addr_width
+
+class IOPort:
+  def __init__(self, name: str, direction: str, width: str):
+    self.name = name
+    self.direction = direction
+    self.width = width
+
+
+def _dfs(node, filter_func):
+  if filter_func(node):
+    yield node
+  for c in node.children():
+    yield from _dfs(c, filter_func)
+
+
+def add_indent(code: List[str]) -> List[str]:
+  return [f'  {line}' for line in code]
+
+
+def parse_ports(ast) -> List[IOPort]:
+
+  io_list = []
+  
+  is_input = lambda node: isinstance(node, Input)
+  is_output = lambda node: isinstance(node, Output)
+  is_io = lambda node: is_input(node) or is_output(node)
+
+  codegen = ASTCodeGenerator()
+  for node in _dfs(ast, is_io):
+    if is_input(node):
+      direction = 'input'
+    else:
+      direction = 'output'
+    width = codegen.visit(node.width) if node.width else ''
+    io_list.append(IOPort(node.name, direction, width))
+
+  return io_list
+
+
+def parse_m_axi_interfaces(top_rtl: str) -> List[AXI]:
```

Comment:
Don't abuse regex when `tapac` already knows [the width of all AXI interfaces](https://github.com/UCLA-VAST/tapa/blob/3ebb143b8499813c45f004e5f7837be9a6358f1f/backend/python/tapa/core.py#L859).